ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -93,115 +57,81 @@ impl From<u64> for OpCode {
         match self {
-            Self::Long(_) => todo!(),
-            Self::Short(op) => op.val(),
+            Self::Local(idx) => idx >> n == 0,
+            Self::ConstantPointer(_) => n >= 64,
         }
     }
-
-    fn is_short(&self) -> bool {
-        matches!(self, Self::Short(_))
-    }
 }
 
 impl fmt::Display for Operand {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self.kind() {
-            OpKind::Local => write!(f, " %{}", self.val())?,
-            OpKind::NotPresent => (),
+        match self {
+            Self::Local(idx) => write!(f, "%{}", idx),
+            _ => todo!(),
         }
-        Ok(())
-    }
-}
-
-#[derive(Debug, PartialEq)]
-pub struct LongOperand(u64);
-
-#[derive(Debug, PartialEq)]
-pub struct ShortOperand(u64);
-
-impl ShortOperand {
-    fn new(kind: OpKind, val: u64) -> ShortOperand {
-        ShortOperand((kind as u64) | (val << SHORT_OPERAND_VALUE_SHIFT))
-    }
-
-    fn kind(&self) -> OpKind {
-        OpKind::from(self.0 & SHORT_OPERAND_KIND_MASK)
-    }
-
-    fn val(&self) -> u64 {
-        self.0 >> SHORT_OPERAND_VALUE_SHIFT
     }
 }
 
-/// An instruction.
-///
-/// An instruction is either a short instruction or a long instruction.
+/// A generic instruction.
 ///
-/// ## Short instruction
+/// This struct is the generic "at-rest" encoding of a specific JIT IR instruction. In order to
+/// access the operands of an instruction, it must first be "interpreted" as the correct specific
+/// instruction.
 ///
-/// - A 64-bit bit-field that encodes the entire instruction inline.
-/// - Can encode up to three short operands.
-/// - Is designed to encode the most commonly encountered instructions.
+/// For example:
 ///
-/// Encoding (LSB first):
 /// ```ignore
-/// field           bit-size
-/// ------------------------
-/// is_short=1      1
-/// opcode          8
-/// short_operand0  18
-/// short_operand1  18
-/// short_operand2  18
-/// reserved        1
+/// let instr: Instruction = LoadInstruction::new(Operand::Local(0));
+/// let loadinstr = LoadInstruction(&instr);
+/// let op = loadinstr.operand();
 /// ```
 ///
-/// Where a short operand is encoded like this (LSB first):
+/// Instructions come in short and long variants. A short instruction packs all of its data inline
+/// into the 64-bit [Instruction], whereas a long instruction is a box pointer to a larger data
+/// structure describing the operands.
+///
+/// The IR is designed such that (where possible) the most common kinds of operations can be
+/// encoded into a short instruction.
+///
+/// Instructions are immutable and must not be copied. If you wish to change an instruction,
+/// you must make a new one and replace the old one. If you wish to copy an instruction, you
+/// should make a new one with the same operands. There are several reasons for this:
+///
+///   1) modifying a (boxed) long instruction could lead to it become short, which may result in
```

Comment:
"lead to it become short" doesn't parse.

Reply:
"lead *it*" would be better.

Reply:
That wouldn't parse either. I assume that "modifying a boxed long instruction could lead to it becoming a short instruction" is what's intended?

Reply:
yes